keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,535 @@
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that have the less number of matching Pods as `maxSkew`.
+
+`minDomains` is an optional parameter and the default value is 0.
+
+### API
+
+New optional parameter called `MinDomains` is introduced to `PodSpec.TopologySpreadConstraint[*]`.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // When the number of domains with matching topology keys is less than `minDomains`,
+  // Pod Topology Spread treats "global minimum" as 0.
+  // As a result,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`,
+  //   scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`,
+  //   scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+  // Default value is 0.
+  // +optional
+  MinDomains *int32
+}
+```
+
+### Implementation details
+
+**Filtering phase**
+
+In Filter of Pod Topology Spread, filtering criteria is
+
+```
+('existing matching num' + 'if self-match (1 or 0)' - 'global min matching num') <= 'maxSkew'
+```
+
+- `existing matching num` denotes the number of current existing matching Pods on the domain.
+- `if self-match` denotes if the labels of Pod is match with selector of the constraint.
+- `global min matching num` denotes the minumun number of matching Pods.
+
+For `whenUnsatisfiable: DoNotSchedule`, Pod Topology Spread treats `global min matching num` as 0
+when the number of domains with matching topology keys is less than `minDomains`.
+
+In `preFilterState`, Pod Topology Spread has `Constraints` field which type is slice of internal `topologySpreadConstraint` struct.
+By introducing `MinDomains` field in `topologySpreadConstraint`,
+Pod Topology Spread can use `MinDomains` to determine the value of `global min matching num` when we calculate filtering criteria.
+
+This change will not increase the complexity of filtering logic and the performance degradation is not expected.
+
+**Scoring phase**
+
+When the number of domains with matching topology keys is less than `minDomains` and `whenUnsatisfiable` equals to `ScheduleAnyway`,
+Pod Topology Spread gives low scores to Nodes on the domains which have the same or more number of matching Pods as `maxSkew`.
+
+In Pod Topology Spread, the higher the score from Score, the lower will be the final score calculated by Normalized Score.
+So, Pod Topology Spread should give low scores to non-preferred Nodes in Score.
+
+In Score, the score for each constraint is evaluated, and the sum of those scores will be a Node score.
+
+Basically, the score for each constraint is evaluated this way:
+
+```
+'existing matching num' * 'topology weight' + 'maxSkew' - 1
```

Comment:
In a way, this formula already gives the semantics that we want.

Ideally, we want to make the formula return a final normalized score of 0 if `existing matching num >= maxSkew` and `topology domains < minDomains` (assuming that there is only one constraint).

Given the way we do normalization, that might not be feasible.

Then we have 2 alternatives:
- Make the violated constraint dominate the score. Return `-1` and use that in Normalization to set a 0 final score.
- Give the violating constraint the highest unnormalized score possible, which is what you are proposing.

I think we might require some experiments with different alternatives and different node/pod configurations to see how the final scores behave. Can you try some? See this PR description as an example:
https://github.com/kubernetes/kubernetes/pull/91918#issue-634940699

Reply:
It might be the case that after your experiments, we just decide to keep the formula and not change anything about scoring. As I said, the formula already gives some of the semantics we want. This feature is more important about `DoNotSchedule` behavior, after all.

Reply:
(Since there are two threads (here and [this thread](https://github.com/kubernetes/enhancements/pull/3030#discussion_r761492956)) of similar topic, I'll reply to them together here.)
> An alternative is to make all the nodes that have existing matching num >= maxSkew (when number of domains is less than minDomains) have the max Score, so that the normalized score is potentially zero.
https://github.com/kubernetes/enhancements/pull/3030#discussion_r761492956

When 
- there are `domainA` and `domainB`,
- and `domainA` has 5 Pods and `domainB` has 10 Pods, 
- and maxSkew = 1,
- and whenUnsatisfiable=ScheduleAnyway
- and the number of domains is less than minDomains,

I think it is better to give a lower final score to domainB (assuming that there is only one constraint). 
With my proposed way, we can give other high scores to each two domains in Score.

> Make the violated constraint dominate the score. Return -1 and use that in Normalization to set a 0 final score.

This way differs from my proposed way in that this way makes the violated constraint dominate the score as you said.

For example: 
| --  |  constraintA  | constraintB | constraintC |
| ---- | ---- | ---- | ---- |
|  nodeA | violate | violate | - |
|  nodeB  | - | violate | violate |
| nodeC | violate | - | - |

violate: "existing matching Pod num on the domain which the node belongs to >= maxSkew" and "the number of domains < minDomains"

In this case, I think it is better to give a higher final score to nodeC than to nodeA and nodeB. 
With my proposed way, Pod Topology Spread gives twice score as usual when existing matching num >= maxSkew. And it is expected that nodeC will get a higher final score than nodeA and nodeB.

For these reasons, I prefer my proposed way.

What do you think about these cases? 

---

> Circling back on what I said earlier, you should prefer not to add code to the KEP. But a not-so-simple formula that has conditionals can be more readable as a code snippet. As long as you are not providing a significant diff, some code is ok in a KEP.
https://github.com/kubernetes/enhancements/pull/3030#discussion_r761492956

Got it :)

--- 

> I think we might require some experiments with different alternatives and different node/pod configurations to see how the final scores behave. Can you try some? See this PR description as an example:

Sure. I'll try it lator 🙇

Reply:
Yeah, your proposed idea is valid.

We need to compare it's behavior with the existing behavior. Do you agree that the existing behavior gives some of the semantics we want?

I'll review again after you share some numbers.

Reply:
> Do you agree that the existing behavior gives some of the semantics we want?

Yes, Pod Topology Spread is already giving a high score to Node with a low number of Pods. 
This is basically the same behavior that `minDomains` want to do.

---

> I think we might require some experiments with different alternatives and different node/pod configurations to see how the final scores behave. Can you try some? 

I examined how the final score changes depending on the multiplier for the violation in various situations.
(violate: "existing matching Pod num on the domain which the node belongs to >= maxSkew" and "the number of domains < minDomains")

**Prerequisite**

Create constraints for two domains (keys: `hostname` and `hostname2`) and both indicate the same as `kubernetes.io/hostname`.
`maxSkew` of `hostname` domain is always 1 and `maxSkew` of `hostname2` domain is 1 or 2.

current resources are:

```
nodes: node1, node2, node3 
- node1 has one pod
- node2 has two pods
- node3 has no pods
(All pods/nodes are same spec.)
```

**current behavior (`minDomains` is not supported) and `maxSkew` of `hostname2` domain is 1**

```
node: node1  TopologyKey: hostname2  score:  1.6094379124341003
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  3.2188758248682006
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final score: [{node1 50} {node2 0} {node3 100}]
```

**current behavior (`minDomains` is not supported) and `maxSkew` of `hostname2` domain is 2**

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  4.218875824868201
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 57} {node2 14} {node3 100}]
```

**the  `minDomains` of `hostname2` domain is 100 and `maxSkew` of `hostname2` domain is 2**

In this case, 
- {the number of domains of  `hostname2`} < { `minDomains` of `hostname2` domain }
- {existing matching Pod num on node2} >= { `maxSkew` of `hostname2` domain }

The score of "node: node2  TopologyKey: hostname2" is affected.

2X when violate:

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  8.437751649736402
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 72}{node2 9}  {node3 100}]
```

3X when violate:

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  12.656627474604603
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [ {node1 80}{node2 6}{node3 100} ]
```

10X when viorate:

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  42.18875824868201
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 93} {node2 2} {node3 100}]
```

**the  `minDomains` of `hostname2` domain is 100 and `maxSkew` of `hostname2` domain is 1**

In this case, 
- {the number of domains of  `hostname2`} < { `minDomains` of `hostname2` domain }
- {existing matching Pod num on node1/node2} >= { `maxSkew` of `hostname2` domain }

The score of "node: node2  TopologyKey: hostname2" and "node: node1  TopologyKey: hostname2" are affected.

2X when violate:

```
node: node1  TopologyKey: hostname2  score:  3.2188758248682006
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  6.437751649736401
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 55} {node2 0} {node3 100}]
```

3X when violate:

```
node: node1  TopologyKey: hostname2  score:  4.828313737302301
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  9.656627474604601
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 50} {node2 0} {node3 100} ]
```

10X when violate:

```
node: node1  TopologyKey: hostname2  score:  16.094379124341003
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  32.18875824868201
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 51} {node2 0} {node3 100}]
```

Reply:
> Do you agree that the existing behavior gives some of the semantics we want?

Yes, Pod Topology Spread is already giving a high score to Node with a low number of Pods. 
This is basically the same behavior that `minDomains` want to do.

---

> I think we might require some experiments with different alternatives and different node/pod configurations to see how the final scores behave. Can you try some? 

I examined how the final score changes depending on the multiplier for the violation in various situations.
(violate: "existing matching Pod num on the domain which the node belongs to >= maxSkew" and "the number of domains < minDomains")

**Prerequisite**

Create constraints for two domains (keys: `hostname` and `hostname2`) and both indicate the same as `kubernetes.io/hostname`.
`maxSkew` of `hostname` domain is always 1 and `maxSkew` of `hostname2` domain is 1 or 2.

current resources are:

```
nodes: node1, node2, node3 
- node1 has one pod
- node2 has two pods
- node3 has no pods
(All pods/nodes are same spec.)
```

**current behavior (`minDomains` is not supported) and `maxSkew` of `hostname2` domain is 1**

```
node: node1  TopologyKey: hostname2  score:  1.6094379124341003
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  3.2188758248682006
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final score: [{node1 50} {node2 0} {node3 100}]
```

**current behavior (`minDomains` is not supported) and `maxSkew` of `hostname2` domain is 2**

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  4.218875824868201
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 57} {node2 14} {node3 100}]
```

**the  `minDomains` of `hostname2` domain is 100 and `maxSkew` of `hostname2` domain is 2**

In this case, 
- {the number of domains of  `hostname2`} < { `minDomains` of `hostname2` domain }
- {existing matching Pod num on node2} >= { `maxSkew` of `hostname2` domain }

The score of "node: node2  TopologyKey: hostname2" is affected.

2X when violate:

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  8.437751649736402
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 72}{node2 9}  {node3 100}]
```

3X when violate:

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  12.656627474604603
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [ {node1 80}{node2 6}{node3 100} ]
```

10X when viorate:

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  42.18875824868201
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  1
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 93} {node2 2} {node3 100}]
```

**the  `minDomains` of `hostname2` domain is 100 and `maxSkew` of `hostname2` domain is 1**

In this case, 
- {the number of domains of  `hostname2`} < { `minDomains` of `hostname2` domain }
- {existing matching Pod num on node1/node2} >= { `maxSkew` of `hostname2` domain }

The score of "node: node2  TopologyKey: hostname2" and "node: node1  TopologyKey: hostname2" are affected.

2X when violate:

```
node: node1  TopologyKey: hostname2  score:  3.2188758248682006
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  6.437751649736401
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 55} {node2 0} {node3 100}]
```

3X when violate:

```
node: node1  TopologyKey: hostname2  score:  4.828313737302301
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  9.656627474604601
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 50} {node2 0} {node3 100} ]
```

10X when violate:

```
node: node1  TopologyKey: hostname2  score:  16.094379124341003
node: node1  TopologyKey: hostname  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  32.18875824868201
node: node2  TopologyKey: hostname  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
node: node3  TopologyKey: hostname  score:  0
final scores:  [{node1 51} {node2 0} {node3 100}]
```

Reply:
It doesn't look like the final scores change much from the existing formula.

Can you get rid of the `hostname` constraint? I don't think it's helping see the difference.

Reply:
> Can you get rid of the hostname constraint? 

Sure. 🙇 

---

**current behavior (minDomains is not supported) and maxSkew of hostname2 domain is 1**

```
node: node1  TopologyKey: hostname2  score:  1.6094379124341003
node: node2  TopologyKey: hostname2  score:  3.2188758248682006
node: node3  TopologyKey: hostname2  score:  0
final scores:  [{node1 66} {node2 0} {node3 100}]
```

**current behavior (minDomains is not supported) and maxSkew of hostname2 domain is 2**

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node2  TopologyKey: hostname2  score:  4.218875824868201
node: node3  TopologyKey: hostname2  score:  1
final scores:  [{node1 75}{node2 25} {node3 100} ]
```

**the minDomains of hostname2 domain is 100 and maxSkew of hostname2 domain is 2**

(The score of "node: node2 TopologyKey: hostname2" is affected.)

2X

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node2  TopologyKey: hostname2  score:  8.437751649736402
node: node3  TopologyKey: hostname2  score:  1
final scores:  [{node1 87}{node2 12} {node3 100} ]
```

3X

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node2  TopologyKey: hostname2  score:  12.656627474604603
node: node3  TopologyKey: hostname2  score:  1
final scores:  [{node1 91} {node2 8} {node3 100}]
```

10X

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node2  TopologyKey: hostname2  score:  42.18875824868201
node: node3  TopologyKey: hostname2  score:  1
final scores:  [{node1 97} {node2 2} {node3 100}]
```


**the minDomains of hostname2 domain is 100 and maxSkew of hostname2 domain is 1**

(The score of "node: node2 TopologyKey: hostname2" and "node: node1 TopologyKey: hostname2" are affected.)

2X

```
node: node1  TopologyKey: hostname2  score:  3.2188758248682006
node: node2  TopologyKey: hostname2  score:  6.437751649736401
node: node3  TopologyKey: hostname2  score:  0
final scores:  [{node1 50} {node2 0} {node3 100}]
```

3X

```
node: node1  TopologyKey: hostname2  score:  4.828313737302301
node: node2  TopologyKey: hostname2  score:  9.656627474604601
node: node3  TopologyKey: hostname2  score:  0
final scores:  [{node1 55} {node2 0} {node3 100}]
```

10X

```
node: node1  TopologyKey: hostname2  score:  16.094379124341003
node: node2  TopologyKey: hostname2  score:  32.18875824868201
node: node3  TopologyKey: hostname2  score:  0
final scores:  [{node1 50} {node2 0}{node3 100} ]
```

Reply:
Why is the non-normalized score of node 1 in `the minDomains of hostname2 domain is 100 and maxSkew of hostname2 domain is 2` affected? Putting a pod there still satisfy the maxSkew, so I would expect it to remain the same as the original algorithm.

Other than that, we can see that the effect of the factor is to reduce the normalized score difference between the nodes under the skew vs the ones over the skew. I think this is good. I think 10X would be too much, but I would like the question above to be answered before commenting further.

Reply:
Also, for `the minDomains of hostname2 domain is 100 and maxSkew of hostname2 domain is 1`, I would expect the normalized score of  node1 to go down as you increase the factor. Could you double check?

Reply:
> Why is the non-normalized score of node 1 in the minDomains of hostname2 domain is 100 and maxSkew of hostname2 domain is 2 affected? 

Mmm, I think it was not affected in the above result. Did I miss something?

**current behavior (minDomains is not supported) and maxSkew of hostname2 domain is 2**

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node2  TopologyKey: hostname2  score:  4.218875824868201
node: node3  TopologyKey: hostname2  score:  1
final scores:  [{node1 75}{node2 25} {node3 100} ]
```

**the minDomains of hostname2 domain is 100 and maxSkew of hostname2 domain is 2**

2X

```
node: node1  TopologyKey: hostname2  score:  2.6094379124341005
node: node2  TopologyKey: hostname2  score:  8.437751649736402
node: node3  TopologyKey: hostname2  score:  1
final scores:  [{node1 87}{node2 12} {node3 100} ]
```

Only node2 score is affected. (4.218875824868201 → 8.437751649736402)

---

> Also, for the minDomains of hostname2 domain is 100 and maxSkew of hostname2 domain is 1, I would expect the normalized score of node1 to go down as you increase the factor. Could you double check?

The results don't seem to be wrong. 

2X

```
node: node1  TopologyKey: hostname2  score:  3.2188758248682006
node: node2  TopologyKey: hostname2  score:  6.437751649736401
node: node3  TopologyKey: hostname2  score:  0
node1:  100 * (6 + 0 - 3) / 6 = 50 
node2:  100 * (6 + 0 - 6) / 6 = 0 
node3:  100 * (6 + 0 - 0) / 6 = 100 
final scores:  [ {node1 50}{node2 0} {node3 100}]
```

3X

```
node: node1  TopologyKey: hostname2  score:  4.828313737302301
node: node2  TopologyKey: hostname2  score:  9.656627474604601
node: node3  TopologyKey: hostname2  score:  0
node1:  100 * (9 + 0 - 4) / 9 = 55 
node2:  100 * (9 + 0 - 9) / 9 = 0 
node3:  100 * (9 + 0 - 0) / 9 = 100 
final scores:  [{node1 55}{node2 0} {node3 100} ]
```

10X

```
node: node1  TopologyKey: hostname2  score:  16.094379124341003
node: node2  TopologyKey: hostname2  score:  32.18875824868201
node: node3  TopologyKey: hostname2  score:  0
node1:  100 * (32 + 0 - 16) / 32 = 50 
node2:  100 * (32 + 0 - 32) / 32 = 0 
node3:  100 * (32 + 0 - 0) / 32 = 100 
final scores:  [{node1 50} {node2 0} {node3 100}]
```

In the above`node: node1  TopologyKey: hostname1  score: XX`, XX is type float64.
https://github.com/kubernetes/kubernetes/blob/489fb9bee3f626b3eeb120a5af89ad8c2b2f1c20/pkg/scheduler/framework/plugins/podtopologyspread/scoring.go#L209

But this score `XX` is converted to `int64` here.
https://github.com/kubernetes/kubernetes/blob/489fb9bee3f626b3eeb120a5af89ad8c2b2f1c20/pkg/scheduler/framework/plugins/podtopologyspread/scoring.go#L212

When normalize, we use this criteria:

```
scores[i].Score = framework.MaxNodeScore * (maxScore + minScore - s) / maxScore
```

https://github.com/kubernetes/kubernetes/blob/489fb9bee3f626b3eeb120a5af89ad8c2b2f1c20/pkg/scheduler/framework/plugins/podtopologyspread/scoring.go#L252

(`s`: the score of the node in the score phase.)

In this experiment, `minScore` is always 0.

```
scores[i].Score = framework.MaxNodeScore * (maxScore + 0 - s) / maxScore
↓
scores[i].Score = framework.MaxNodeScore * (maxScore - s) / maxScore
↓
scores[i].Score = framework.MaxNodeScore - s / maxScore
```

`s / maxScore` for node3 is always 100 because `s == maxScore`
`s / maxScore` for node1 **should be** always 50 because Node3's original score (in type float64) is twice as high as Node1's original score (in type float64). But, we convert scores for each Node to type `int` in score phase, so, the score for node1 is `55` when 3X.

Reply:
> Mmm, I think it was not affected in the above result. Did I miss something?

You are right.... not sure what I was looking at.

Thanks for the extra analysis on the int/float conversions.

I would say a 2X factor is enough to strengthen the differentiation between valid and violating nodes. If you agree, could you update the doc?

Reply:
> I would say a 2X factor is enough to strengthen the differentiation between valid and violating nodes. If you agree, could you update the doc?

Yes, agree with that. The document has already been written about it as 2X. So, I only fixed the wording of some points.